src/database.rs

```diff
@@ -125,6 +125,82 @@ impl Database {
         Ok(Self { pool })
     }
 
+    pub async fn insert_pending_identity(
+        &self,
+        group_id: usize,
+        identity: &Hash,
+    ) -> Result<(), Error> {
+        self.pool
+            .execute(
+                sqlx::query(
+                    r#"INSERT INTO pending_identities (group_id, commitment)
+                           VALUES ($1, $2);"#,
+                )
+                .bind(group_id as i64)
+                .bind(identity),
+            )
+            .await?;
+        Ok(())
+    }
+
+    pub async fn mark_identity_inserted(
+        &self,
+        group_id: usize,
+        commitment: &Hash,
+        block_number: usize,
+        tree_idx: usize,
+    ) -> Result<(), Error> {
+        self.pool
+            .execute(
+                sqlx::query(
+                    r#"UPDATE pending_identities
+                           SET mined_in_block = $1, assigned_leaf_idx = $2
+                           WHERE group_id = $3 AND commitment = $4;"#,
+                )
+                .bind(block_number as i64)
+                .bind(tree_idx as i64)
+                .bind(group_id as i64)
+                .bind(commitment),
+            )
+            .await?;
+        Ok(())
+    }
+
+    pub async fn pending_identity_exists(
+        &self,
+        group_id: usize,
+        identity: &Hash,
+    ) -> Result<bool, Error> {
+        let row = self
+            .pool
+            .fetch_optional(
+                sqlx::query(
+                    r#"SELECT 1
+                           FROM pending_identities
+                           WHERE group_id = $1 AND commitment = $2
+                           LIMIT 1;"#,
+                )
+                .bind(group_id as i64)
+                .bind(identity),
+            )
+            .await?;
+        Ok(row.is_some())
+    }
+
+    pub async fn get_oldest_unprocessed_identity(&self) -> Result<Option<(usize, Hash)>, Error> {
+        let row = self
+            .pool
+            .fetch_optional(sqlx::query(
+                r#"SELECT group_id, commitment
+                           FROM pending_identities
+                           WHERE mined_in_block IS NULL
+                           ORDER BY created_at ASC
+                           LIMIT 1;"#,
+            ))
+            .await?;
+        Ok(row.map(|row| (usize::try_from(row.get::<i64, _>(0)).unwrap(), row.get(1))))
```

Comment:
It's a bit annoying that there is no `usize` impl, but understandable since some databases (i.e. postgres) don't have a `u64` type.

FWIW, this is a bit more concise `row.get::<i64, _>(0).try_into().unwrap()`